Scroll To Text Fragment Text Directives don't find text with additional unrendered white space in their node data.

megangardner opened the pull request:
#### ddae86e7776380e5f787464d595a6d44dac3cf1b
<pre>
Scroll To Text Fragment Text Directives don&apos;t find text with additional unrendered white space in their node data.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267460">https://bugs.webkit.org/show_bug.cgi?id=267460</a>
<a href="https://rdar.apple.com/120913588">rdar://120913588</a>

Reviewed by Ryosuke Niwa.

The spec for Scroll to Text Fragment says to just look a the .data() on the node
to determine it&apos;s text content and search for the fragment directive in the page
text. However, this does not account for extra white space that is removed when rendering.
Therefore, we should use TextItterator which does account for a remove this unrendered white
space. This is technically against the spec, but other engines similarly ignore this
extraneous white space when searching, so the spec should be updated accordingly.
If TextItterator does not find the text, we fall back to the original spec way of finding
the text.

* LayoutTests/http/tests/scroll-to-text-fragment/whitespace-collapse-expected.html: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/whitespace-collapse.html: Added.
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::findRangeFromNodeList):

Canonical link: <a href="https://commits.webkit.org/273016@main">https://commits.webkit.org/273016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5fd98803c26a587916202c019b250b9a58033c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12721 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35891 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36561 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30753 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9867 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29803 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10753 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9372 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9473 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37869 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30820 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30603 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35583 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9601 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7541 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33479 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11393 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29907 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7820 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10173 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
LGTM

---

Editorial comment: TextIterator typo and a grammar fix for this line?

> Therefore, we should use TextItterator which does account for a remove this unrendered white
space.

(Not part of this PR, but maybe that layout test should be WPT-ified when this issue is raised to modify the spec!)

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -151,12 +151,24 @@ static std::optional<SimpleRange> findRangeFromNodeList(const String& query, con
+
+    if (!foundText.collapsed())
+        return foundText;
+
```
So we fallback to just matching text when there is no match?
We should probably mention that in the commit message.

megangardner replied:
Yes, and good point!

megangardner committed:
Scroll To Text Fragment Text Directives don't find text with additional unrendered white space in their node data.
https://bugs.webkit.org/show_bug.cgi?id=267460
rdar://120913588

Reviewed by Ryosuke Niwa.

The spec for Scroll to Text Fragment says to just look a the .data() on the node
to determine it's text content and search for the fragment directive in the page
text. However, this does not account for extra white space that is removed when rendering.
Therefore, we should use TextItterator which does account for a remove this unrendered white
space. This is technically against the spec, but other engines similarly ignore this
extraneous white space when searching, so the spec should be updated accordingly.
If TextItterator does not find the text, we fall back to the original spec way of finding
the text.

* LayoutTests/http/tests/scroll-to-text-fragment/whitespace-collapse-expected.html: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/whitespace-collapse.html: Added.
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::findRangeFromNodeList):

Canonical link: https://commits.webkit.org/273016@main
